README.md
adde
====

Arduino Development and Debug Environment.  
Used to debug issues of cross platform Arduino software projects in a standard PC environment using
the Arduino device as a proxy.

Features
========

- fully Arduino API conform
- encourages cross platform development
- full connection to the real Arduino periphery
- no need to re-flash the Arduino device during development to speed up compile/test cycles and preserve flash life-time
- rich debug and log functionality (user log entries via the `addeLog()` function)
- space is only limited by the development PC (not by the MCU)
- helps you find issues like out-of-bound, use-after-free, stack overflow, heap overflow, memory leaks and more
- <1ms remote command latency is possible
- enables the use of tools like GDB, Dr.Memory and Valgrind
- Linux and Windows are supported

Not included is support for MCU specific functions (like special registers or ASM commands) as well
as MCU or peripheral hardware emulation. Those features may be found in [simavr](https://github.com/buserror/simavr).

Usage
=====

Compile and flash the [adde firmware](src/arduino) with the Arduino IDE to your device.  
Copy your .ino and related files to the usr directory and execute `make clean all`.  
Run bin/adde with the serial port of your Arduino device.

Example
=======

Adde comes with a simple blink example in [usr](usr). Just compile everything with GCC using the provided make file

    make clean all

and run, e.g. on Windows, adde.exe (assuming Arduino is connected via COM1):

    bin/adde.exe COM1

Standard input and output will act as interface of Serial. That means if you read or write in Arduino from or to  
Serial it will be from or to the Windows console.  
Run `bin/adde.exe --help` for a list of options.

Building
========

The following dependencies are given:  
- C99
- C++11

Edit Makefile to match your target system configuration.  
Building the program:  

    make

Add one of the following flags to make if needed:
- NO_GPL=1
- UNICODE=1
- DEBUG=1
- USR=usr/

[![Linux GCC Build Status](https://img.shields.io/travis/daniel-starke/adde/master.svg?label=Linux)](https://travis-ci.org/daniel-starke/adde)
[![Windows Visual Studio Build Status](https://img.shields.io/appveyor/ci/danielstarke/adde/master.svg?label=Windows)](https://ci.appveyor.com/project/danielstarke/adde)    

Files
=====

|Name            |Meaning
|----------------|--------------------------------------------------
|**src/arduino/**|**Arduino device firmware for adde.**
|arduino.*       |Main implementation of the device functionality.
|Crc16.hpp       |CRC16 implementation according to [RFC 1662](https://tools.ietf.org/html/rfc1662).
|Framing.hpp     |Framing functions for the communication protocol.
|Get*            |Functions to provide Arduino specific constants.
|Meta.hpp        |C++11 meta programming helpers (derived from STL).
|New.cpp         |C++ placement new implementation.
|Optional.hpp    |Static allocated optional object creation helper.
|Pre.hpp         |C/C++ pre-processor helpers.
|Protocol.hpp    |Adde protocol specific constants.
|Reset.cpp       |Soft device reset implementation.
|Target.hpp      |Arduino target specific functions and macros.
|Utility.hpp     |Adde protocol utility functions.
|                |
|**src/utility/**|**Utility functions for the Arduino API.**
|adde.hpp        |Adde function return value helpers.
|argp*, getopt*  |Command-line parser.
|cvutf8.*        |UTF-8 conversion functions.
|mingw-unicode.h |Unicode enabled main() for MinGW targets.
|serial.*        |Serial interface functions.
|target.h        |Target specific functions and macros.
|tchar.*         |Functions to simplify ASCII/Unicode support.
|                |
|**src/**        |**Arduino API implementation.**
|abi.cpp         |C++ ABI specific functions.
|Arduino.*       |Arduino API implementation (most of it).
|binary.h        |Arduino binary constants.
|ConsoleSerial.h |Console interface for the Serial class.
|EEPROM.h        |Arduino EEPROM class interface.
|HardwareSerial.h|Arduino HardwareSerial class interface.
|License.i       |ADDE license text include.
|LiquidCrystal.h |Arduino LiquidCrystal class interface.
|new.*           |C++ `new` specific functions.
|pins_arduino.h  |Arduino PIN definitions.
|Print.*         |Arduino Print class and interface.
|Printable.h     |Arduino Printable class interface.
|SoftwareSerial.h|Arduino SoftwareSerial class interface.
|SPI.h           |Arduino SPI class interface.
|Stream.*        |Arduino Stream class and interface.
|WCharacter.*    |Arduino character specific functions.
|Wire.h          |Arduino Wire (I2C) class interface.
|WMath.cpp       |Arduino math functions.
|WString.*       |Arduino String class and interface.

Limitations
===========

- Callbacks are unimplemented (and so are `Wire.onReceive` and `Wire.onRequest`).
- `String::replace`, `String::remove`, `Stream::findUntil` and `Stream::findMulti` require GPL code and are excluded by default (see NO_GPL in [Makefile](Makefile))
- Remote functions returning pointers or modifying passed pointers or variables are not supported.
- The implemented soft reset executed at application startup does not resets all states on the remote device (see [`softReset()`](src/arduino/Reset.cpp).
- Primitive data types are handled with the same range as on the target device. However, implicit type promotion may create different results. This could be corrected with incorporation of `boost::safe_numerics` in future releases.

License
=======

See [copying file](doc/COPYING).  

Contributions
=============

No content contributions are accepted. Please file a bug report or feature request instead.  
This decision was made in consideration of the used license.


doc/Changelog.txt
version numbering schema:
M.N.P
| | +-- patch: increased if only bug fixed were incorporated
| +---- minor: increased if command-line syntax/semantic breaking changes were applied
+------ major: increased if elementary changes (from user's point of view) were made

1.0.2 (2020-08-31)
 - changed: Utility.hpp to use make_index_sequence and index_sequence from Meta.hpp
 - fixed: assignment/copy-constructor mismatch in Crc16
 - fixed: wrong handling of broken quotation in Framing
 - fixed: shortTimeout type in _FVI::waitForResult()
 - fixed: declaration/definition scope of handleResult()
 - fixed: maybe-uninitialized warning for result.res in CallResult
 - fixed: unused function argument res in callUnpacked()
 - fixed: implicit casts from size_t to uint8_t in arduino.ino

1.0.1 (2019-06-28)
 - fixed: virtual COM port removal detection on Windows
 - fixed: high CPU usage for serial read on Windows
 - fixed: string output in Unicode mode on Windows (due to __USE_MINGW_ANSI_STDIO)

1.0.0 (2019-05-30)
 - first release


src/Arduino.h
/**
 * @file Arduino.h
 * @author Daniel Starke
 * @copyright Copyright 2019 Daniel Starke
 * @date 2019-03-08
 * @version 2020-05-16
 */
#ifndef __ADDE_ARDUINO_H__
#define __ADDE_ARDUINO_H__

#include <inttypes.h>
#include <math.h>
#include <stdbool.h>
#include <stdint.h>
#include <stdlib.h>
#include <string.h>
#include "utility/adde.hpp"
#include "binary.h"
#include "pins_arduino.h"


#define PI 3.1415926535897932384626433832795
#define HALF_PI 1.5707963267948966192313216916398
#define TWO_PI 6.283185307179586476925286766559
#define DEG_TO_RAD 0.017453292519943295769236907684886
#define RAD_TO_DEG 57.295779513082320876798154814105
#define EULER 2.718281828459045235360287471352


#undef abs
#define min(x, y) (((x) < (y)) ? (x) : (y))
#define max(x, y) (((x) > (y)) ? (x) : (y))
#define abs(x) (((x) > 0) ? (x) : -(x))
#define constrain(x, low, high) (((x) < (low)) ? (low) : (((x) > (high)) ? (high) : (x)))
#define round(x) (((x) >= 0) ? long((x) + 0.5) : long((x) - 0.5))
#define radians(deg) ((deg) * DEG_TO_RAD)
#define degrees(rad) ((rad) * RAD_TO_DEG)
#define sq(x) ((x) * (x))
#undef RANDOM_MAX
#define RANDOM_MAX 0x7FFFFFFF


#define HIGH _HIGH
#define LOW _LOW
#define INPUT _INPUT
#define INPUT_PULLUP _INPUT_PULLUP
#define OUTPUT _OUTPUT
#define SERIAL _SERIAL
#define DISPLAY _DISPLAY
#define LSBFIRST _LSBFIRST
#define MSBFIRST _MSBFIRST
#define CHANGE _CHANGE
#define FALLING _FALLING
#define RISING _RISING
#define DEFAULT _DEFAULT
#define EXTERNAL _EXTERNAL
#define INTERNAL1V1 _INTERNAL1V1
#define INTERNAL _INTERNAL
#define INTERNAL2V56 _INTERNAL2V56
#define INTERNAL2V56_EXTCAP _INTERNAL2V56_EXTCAP
#define F_CPU _F_CPU


#define interrupts() _INTERRUPTS = true
#define noInterrupts() _INTERRUPTS = false


#define PSTR(...) __VA_ARGS__


#define clockCyclesPerMicrosecond() (F_CPU / 1000000L)
#define clockCyclesToMicroseconds(x) ((x) / clockCyclesPerMicrosecond())
#define microsecondsToClockCycles(x) ((x) * clockCyclesPerMicrosecond())


#define lowByte(x) ((uint8_t)((x) & 0xFF))
#define highByte(x) ((uint8_t)((x) >> 8))


#define bit(x) (1UL << (x))
#define bitRead(x, bit) (((x) >> (bit)) & 0x01)
#define bitSet(x, bit) ((x) = (__typeof__(x))((x) | (1UL << (bit))))
#define bitClear(x, bit) ((x) = (__typeof__(x))((x) & (~(1UL << (bit)))))
#define bitWrite(x, bit, val) ((val) ? bitSet((x), (bit)) : bitClear((x), (bit)))


#ifndef _NOP
#define _NOP() do { __asm__ volatile ("nop"); } while (0)
#endif


typedef uint8_t byte;
typedef uint16_t word;


#ifdef __cplusplus
extern "C" {
#endif


extern volatile bool _INTERRUPTS;
extern uint8_t _HIGH;
extern uint8_t _LOW;
extern uint8_t _INPUT;
extern uint8_t _INPUT_PULLUP;
extern uint8_t _OUTPUT;
extern uint8_t _SERIAL;
extern uint8_t _DISPLAY;
extern uint8_t _LSBFIRST;
extern uint8_t _MSBFIRST;
extern uint8_t _CHANGE;
extern uint8_t _FALLING;
extern uint8_t _RISING;
extern uint8_t _DEFAULT;
extern uint8_t _EXTERNAL;
extern uint8_t _INTERNAL1V1;
extern uint8_t _INTERNAL;
extern uint8_t _INTERNAL2V56;
extern uint8_t _INTERNAL2V56_EXTCAP;
extern uint32_t _F_CPU;


#ifndef NOT_AN_INTERRUPT
#define NOT_AN_INTERRUPT (-1)
#endif /* NOT_AN_INTERRUPT */


#ifndef INVALID_PIN
/** Protocol specific and may not be changed. */
#define INVALID_PIN (0xFF)
#endif /* INVALID_PIN */


#ifndef INVALID_PORT
/** Protocol specific and may not be changed. */
#define INVALID_PORT (0xFF)
#endif /* INVALID_PORT */


#ifndef INVALID_REFERENCE
/** Protocol specific and may not be changed. */
#define INVALID_REFERENCE (0xFF)
#endif /* INVALID_REFERENCE */


#ifndef INVALID_SERIAL_MODE
/** Protocol specific and may not be changed. */
#define INVALID_SERIAL_MODE (0xFF)
#endif /* INVALID_SERIAL_MODE */


#ifndef INVALID_SPI_CLOCK
/** Protocol specific and may not be changed. */
#define INVALID_SPI_CLOCK (0xFF)
#endif /* INVALID_SPI_CLOCK */


#ifndef INVALID_SPI_MODE
/** Protocol specific and may not be changed. */
#define INVALID_SPI_MODE (0xFF)
#endif /* INVALID_SPI_MODE */


void yield(void);

void init(void);
void initVariant(void);

int _atexit(void (* func)()) __attribute__((weak));
#define atexit _atexit

void pinMode(uint8_t pin, uint8_t mode);
void digitalWrite(uint8_t pin, uint8_t val);
int digitalRead(uint8_t pin);
int analogRead(uint8_t pin);
void analogReference(uint8_t mode);
void analogWrite(uint8_t pin, int val);
void analogReadResolution(int val);
void analogWriteResolution(int val);

unsigned long millis(void);
unsigned long micros(void);
void delay(unsigned long val);
void delayMicroseconds(unsigned int val);
unsigned long pulseIn(uint8_t pin, uint8_t val, unsigned long timeout);
unsigned long pulseInLong(uint8_t pin, uint8_t val, unsigned long timeout);

void shiftOut(uint8_t dataPin, uint8_t clockPin, uint8_t bitOrder, uint8_t val);
uint8_t shiftIn(uint8_t dataPin, uint8_t clockPin, uint8_t bitOrder);

void attachInterrupt(uint8_t interrupt, void (* isr)(void), int mode);
void detachInterrupt(uint8_t interrupt);

void setup(void);
void loop(void);


/* non-Arduino helper function for debugging (works like printf) */
#if defined(__GNUC__)
int addeLog(const char * fmt, ...) __attribute__((format(gnu_printf, 1, 2)));
#else /* !GCC */
int addeLog(const char * fmt, ...);
#endif /* GCC */


#ifdef __cplusplus
} /* extern "C" */
#endif


#ifdef __cplusplus


#include <new>
#include "WCharacter.h"
#include "WString.h"
#include "ConsoleSerial.h"
#include "HardwareSerial.h"


typedef bool boolean;


ADDE_F(_int) _digitalRead(uint8_t pin);
#define digitalRead _digitalRead
ADDE_F(_int) _analogRead(uint8_t pin);
#define analogRead _analogRead

unsigned long pulseIn(uint8_t pin, uint8_t val, unsigned long timeout = 1000000UL);
ADDE_F(_unsigned_long) _pulseIn(uint8_t pin, uint8_t val, unsigned long timeout = 1000000UL);
#define pulseIn _pulseIn
unsigned long pulseInLong(uint8_t pin, uint8_t val, unsigned long timeout = 1000000UL);
ADDE_F(_unsigned_long) _pulseInLong(uint8_t pin, uint8_t val, unsigned long timeout = 1000000UL);
#define pulseInLong _pulseInLong

ADDE_F(_uint8_t) _shiftIn(uint8_t dataPin, uint8_t clockPin, uint8_t bitOrder);
#define shiftIn _shiftIn

void tone(uint8_t pin, unsigned int frequency, unsigned long duration = 0);
void noTone(uint8_t pin);

long random(long _max);
long random(long _min, long _max);
void randomSeed(unsigned long seed);
long map(long val, long fromLow, long fromHigh, long toLow, long toHigh);

uint16_t makeWord(uint16_t w);
uint16_t makeWord(uint8_t h, uint8_t l);
#define word(...) makeWord(__VA_ARGS__)


#endif /* __cplusplus */


#endif /* __ADDE_ARDUINO_H__ */
